Runs: Rewrite Runs section (#2011)
wandb/docs: models/runs/forking.mdx

---
description: Explore different parameters or models from a specific point in an experiment without impacting the original run.
title: Fork a run
---

<Warning>
The ability to fork a run is in private preview. Contact W&B Support at support@wandb.com to request access to this feature.
</Warning>

> Explore different parameters or models from a specific point in an experiment without impacting the original run. To do this, "fork" from an existing W&B run. When you fork from a run, W&B creates a new run based on the the source run's `run ID` and `step`. After you initialize a forked run, you can continue logging to the new run independently from the source run.

Reviewer: Not convinced about "fork". It's what we are calling it, I think we can avoid the quotes.

{/* Use `fork_from` when you initialize a run with [`wandb.init()`](/models/ref/python/functions/init) to "fork" from an existing W&B run. When you fork from a run, W&B creates a new run using the `run ID` and `step` of the source run. */}

<Note>

> * Forking a run requires [`wandb`](https://pypi.org/project/wandb/) SDK version >= 0.16.5

Reviewer: This is a very old SDK version at this point, can we lose this requirement?

> * Forking a run requires monotonically increasing steps. You can not use non-monotonic steps defined with [`define_metric()`](/models/ref/python/experiments/run#define_metric) to set a fork point because it would disrupt the essential chronological order of run history and system metrics.

Reviewer: ```suggestion
* Forking a run requires monotonically increasing steps. Attempting to set a fork point with non-monotonic steps defined with [`define_metric()`](/models/ref/python/experiments/run#define_metric) is not supported, because it would disrupt the essential chronological order of run history and system metrics.
```

</Note>

> Use the `fork_from` parameter in [`wandb.init()`](/models/ref/python/functions/init) to fork from an existing run. Specify the source run's unique `run ID` and the `step` you want to start the forked run from. 

Reviewer: Curious why it's showing here but commented out in line 12?

<Note>
See [Unique run identifiers](/models/runs/run-identifiers#unique-run-identifiers) to learn more about run IDs and how to locate them.
</Note>

## Fork from a previously logged run

> The following code snippet demonstrates how to fork from a run that you previously logged to W&B. This might occur if you want to fork from a run in a different script or notebook. Or if someone else logged the source run that you want to fork from.

Reviewer: ```suggestion
The following code snippet shows how to fork from a run that was previously logged to W&B. Use this approach to fork from a run in a different script or notebook, or from a run logged by a different user or service account.
```

Replace `<source-run-id>`, `<project>`, and `<entity>` with your own values.

```python
import wandb

# The unique ID of the source run to fork from
source_run_id = "<source-run-id>"

# Specify the step to fork from
fork_step = 200

# Fork the run
with wandb.init(
    project="<project>",
    entity="<entity>",
    fork_from=f"{source_run_id}?_step={fork_step}",
) as forked_run:
    pass
```

## Fork from a run in the same script

The following code snippet shows how to fork from a run within the same script. This might occur if you want explore different parameters or models from a specific point within the same script or notebook.

```python
import wandb

# Initialize a run
with wandb.init(
    project="<project>",
    entity="<entity>"
) as original_run:
    # ...training logic goes here ...
    pass

# Specify the step to fork from
fork_step = 200

# Use the original run's ID and specify the step to fork from
with wandb.init(
    project="<project>",
    entity="<entity>",
    fork_from=f"{original_run.id}?_step={fork_step}",
) as forked_run:
    # ...training logic goes here ...
    pass
```

> Note that you use the `original_run.id` property to obtain the unique run ID of the original run.

Reviewer: ```suggestion
Use the `original_run.id` property to obtain the unique run ID of the original run.
```

> ## Example script

Reviewer: ```suggestion
### Example script
```

The ability to fork a run is in private preview. Contact W&B Support at support@wandb.com to request access to this feature.
</Warning>

> Explore different parameters or models from a specific point in an experiment without impacting the original run. To do this, *fork* from an existing W&B run. When you fork from a run, W&B creates a new run based on the the source run's `run ID` and `step`. After you initialize a forked run, you can continue logging to the new run independently from the source run.

Reviewer: What about "To explore....., *fork* from an existing...."

<Note>
* Forking a run requires [`wandb`](https://pypi.org/project/wandb/) SDK version >= 0.16.5

> * Forking a run requires monotonically increasing steps. Attempting to set a fork point with non-monotonic steps defined with [`define_metric()`](/models/ref/python/experiments/run#define_metric) is not supported, because it would disrupt the essential chronological order of run history and system metrics.

Reviewer: Nit: Consider replacing some of these gerunds.
